text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+127.0.0.1 - - [13/Aug/2020 22:47:38] "GET /.git/objects/5c/29792762ac54c0ce7614621c48f88eb5fb0352 HTTP/1.1" 200 -
+...
+```
+
+Since the structure is simple and we know the URL for every object if we know its hash, it is straightforward to fetch a single object from the server. Other transports are a bit more sophisticated, but the core idea is similar.
+
+### Pack files
+
+We do notice that our repo has quite a few objects:
+
+```console
+$ find .git/objects/ -type f | wc -l
+24
+```
+
+We can run garbage collection on these objects, causing git to remove unreachable ones and "pack" the others:
+
+```console
+$ git gc
+...
+$ find .git/objects/ -type f | wc -l
+4
+```
+
+Where did our files go? They are in a pack file:
+
+```console
+$ git verify-pack --verbose ./.git/objects/pack/pack-80a4d68ade4dcac5913a0e825ff50c6cb968c5a8.idx
+23d9244c918044b4f1e28de295065b41e029f37c commit 1122 848 12
+ff51a7fc40b7fb6f165d3a92ab88d8d2bf8c0b18 commit 1117 844 860
+...
+```
+
+Pack files are archives that contain multiple objects. They make things a little easier on the filesystem and the HTTP transfers, but more importantly delta compression is applied when files are in a pack.
+
+### Git summary
+
+Git stores different types of objects indexed by hashes, compressed in a very simple key-value scheme on the local filesystem. Every object contains a header indicating its type (commit, tree, blob). These objects can be fetched by hash even across the network. Parent relationships are indicated by hash as well.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+* What is the actual impact on the execution engine?
+* Do we serialize hashes or contents?
+* Do we store DAG nodes statically or dynamically (`Box<dyn ...>`)?
+
+All these questions should be answered before this RFC is merged.
+
+## Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+While not covered in detail in this RFC, these changes lay the groundwork for exciting future possiblities:
+
+### CDNs
+
+Since retrieving objects is just a matter of fetching by hash, we can leverage a CDN, for example via HTTP. Whenever a node needs an object `435abcf`, it can try to `GET http://casper.cdn.net/objects/435abcf` first. If found, we just saved the network the burden of sending the missing piece. This is not limited to the "official" CDN, validators are free to run one themselves, we may even add mechanisms to discover them.
+
+Open questions with this approach is who puts data on the CDN in the first place:
+
+* An easy first take is to upload daily snapshots of the whole chain ourselves. This already considerably reduces the burden on the network when new nodes are joining.
+* Another approach is to allow any node to push objects to the CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus specialized software.
```

Comment:
```suggestion
* An easy first implementation would be an opt-in (via config.toml) behavior for a node instance to upload daily snapshots of the entire chain to a target CDN, and an opt-in behavior to use one or more CDN's. CasperLabs operated nodes would operate a CDN and opt-in to both options using our own CDN. Node operators who trust use could opt-in to use the CDN we provide. Those who do not are free to wire up their own CDN if they like. 
** This minimal version would considerably reduce the burden on the network of new nodes joining, at the cost of periodic activity spikes on nodes pushing snap shots to a CDN.
* Another approach is to allow any node to push objects to a CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus would need some specialized software to enable.
```

Reply:
Manually added in 463121f